src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,582 @@
+            Mask values selected in ``[0, 1]``:
+            ``1`` for tokens that are NOT MASKED, ``0`` for MASKED tokens.
+            `What are attention masks? <../glossary.html#attention-mask>`__
+        token_type_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, input_ids_length)`, `optional`, defaults to :obj:`None`):
+            Segment token indices 
+        position_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`, defaults to :obj:`None`):
+            Indices of positions of each input sequence tokens in the position embeddings.
+            Selected in the range ``[0, config.max_position_embeddings - 1]``.
+            `What are position IDs? <../glossary.html#position-ids>`_
+       
+"""
+
+
+
+class BlenbderbotGeneratorModel(BlenderbotPretrained):
+    """
+    implements a blenderbot generator model with an encoder and decoder
+    """
+    def __init__(self, config):
+        self.pad_idx = config.pad_idx
+        self.start_idx = config.start_idx
+        self.end_idx = config.end_idx
+        super().__init__(config)
+        self.embeddings = nn.Embedding(config.vocab_size, config.embedding_size, self.pad_idx)
+        self.n_positions = config.n_positions if config.n_positions > 0 else 1024
+        self.encoder = BlenderbotEncoder(
+            n_heads=config.n_heads,
+            n_layers=config.n_layers,
+            embedding_size=config.embedding_size,
+            ffn_size=config.ffn_size,
+            vocabulary_size=config.vocab_size,
+            embedding=self.embeddings,
+            dropout=config.dropout,
+            relu_dropout=config.relu_dropout,
+            attention_dropout=config.attention_dropout,
+            padding_idx=self.pad_idx,
+            learn_positional_embeddings=config.learn_positional_embeddings,
+            n_positions=self.n_positions,
+            activation=config.activation,
+        )
+        self.decoder = BlenderbotDecoder(
+            n_heads=config.n_heads,
+            embedding_size=config.embedding_size,
+            ffn_size=config.ffn_size,
+            n_layers=config.n_layers,
+            vocabulary_size=config.vocab_size,
+            embedding=self.embeddings,
+            attention_dropout=config.attention_dropout,
+            dropout=config.dropout,
+            relu_dropout=config.relu_dropout,
+            learn_positional_embeddings=config.learn_positional_embeddings,
+            padding_idx=self.pad_idx,
+            n_positions=self.n_positions,
+            activation=config.activation
+        )
+        self.init_weights()
+
+    def forward(self, input_ids, decoder_input_ids=None, position_ids=None, token_type_ids=None):
+        if input_ids is None:
+            raise ValueError("Your input tensor is None, you should specify it")
+        encoder_out = self.encoder(input_ids, token_type_ids)
+        assert isinstance(encoder_out, tuple)
+
+        def prepare_decoder_inputs(input_ids):
```

Comment:
What does this function do exactly?

Reply:
This shift the input decoder one token to the right and wrap the last non pad token. Same as in the bart model.
I should change the `0` value in `index_of_eos = (input_ids.ne(0).sum(dim=1) - 1).unsqueeze(-1)` into `pad_token_id`